File: dart/test/logical_helper_contract_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:linkedspec_dart/linkedspec_dart.dart';
import 'package:linkedspec_dart/src/cli/primary_cli.dart';
import 'package:linkedspec_dart/src/runtime/interpreter.dart'
    show runtimeLogicalTruth;
import 'package:test/test.dart';

Map<String, Object?> _contract() {
  return (jsonDecode(
            File(
              '../capability_conformance/logical_helper_contract.json',
            ).readAsStringSync(),
          )
          as Map)
      .cast<String, Object?>();
}

Map<String, Object?> _fixture(Map<String, Object?> contract, String id) {
  final fixtures = (contract['fixtures']! as Map).cast<String, Object?>();
  return (fixtures[id]! as Map).cast<String, Object?>();
}

List<Map<String, Object?>> _invalidCases(Map<String, Object?> contract) {
  return (contract['invalid_arity_cases']! as List)
      .cast<Map<Object?, Object?>>()
      .map((row) => row.cast<String, Object?>())
      .toList(growable: false);
}

List<Map<String, Object?>> _invalidFixtures(Map<String, Object?> contract) {
  final fixtures = (contract['fixtures']! as Map).cast<String, Object?>();
  return (fixtures['invalid_arity']! as List)
      .cast<Map<Object?, Object?>>()
      .map((row) => row.cast<String, Object?>())
      .toList(growable: false);
}

CompiledSpec _compileSource(String source) {
  final parsed = parseSpecWithStagedUserFunctionDefinitions(source);
  validateSpec(parsed);
  return compileSpec(parsed);
}

CompiledSpec _reconstructFromEmittedPayload(
  CompiledSpec compiled,
  String identity,
) {
  final emitted = emitDartSourceV1(compiled, identity);
  final encoded = RegExp(
    "const _compiledSpecJsonBase64 = '([^']+)';",
  ).firstMatch(emitted)![1]!;
  final normalized = (jsonDecode(utf8.decode(base64Decode(encoded))) as Map)
      .cast<String, Object?>();
  return compileSpec(SpecFile.fromJson(normalized));
}

Map<String, Object?> _caseFor(List<Map<String, Object?>> cases, String id) {
  return cases.singleWhere((row) => row['id'] == id);
}

Object? _runtimeValue(Map<String, Object?> value) {
  return switch (value['kind']) {
    'null' => null,
    'boolean' || 'number' || 'string' => value['value'],
    'array' => [
      for (final item in value['items']! as List)
        _runtimeValue((item as Map).cast<String, Object?>()),
    ],
    'harray' => {
      for (final entry
          in (value['entries']! as Map).cast<String, Object?>().entries)
        entry.key: _runtimeValue((entry.value! as Map).cast<String, Object?>()),
    },
    'codeblock' => parseActionBlock('fail("codeblock must not run")'),
    final kind => throw StateError('unsupported logical value kind $kind'),
  };
}

void _expectArityDiagnostic(
  RuntimeInterpreterException error,
  Map<String, Object?> row,
) {
  final diagnostic = error.diagnostic?.toJson();
  expect(diagnostic, isNotNull, reason: row['id']! as String);
  expect(diagnostic!['stage'], 'helper_arity_mismatch');
  expect(diagnostic['code'], row['expected_code']);
  expect(diagnostic['helper_name'], row['helper_name']);
  expect(diagnostic['actual_arity'], row['actual_arity']);
  expect(diagnostic['expected_arity'], row['expected_arity']);
  expect(diagnostic['rule_label'], 'Top');
  expect(error.message, contains('helper_arity_mismatch'));
  expect(error.message, contains('helper_name=${row['helper_name']}'));
  expect(error.message, contains('actual_arity=${row['actual_arity']}'));
  expect(error.message, isNot(contains('must not run')));
}

void _expectGeneratedArityDiagnostic(
  GeneratedSourceException error,
  Map<String, Object?> row,
  String identity,
) {
  final diagnostic = error.toJson();
  expect(diagnostic['stage'], 'execute_generated');
  expect(diagnostic['code'], 'generated_execution_failed');
  expect(diagnostic['source_identity'], identity);
  expect(diagnostic['rule_label'], 'Top');
  expect(diagnostic['handler_family'], 'default');
  expect(diagnostic['detail'], contains('helper_arity_mismatch'));
  expect(diagnostic['detail'], contains('helper_name=${row['helper_name']}'));
  expect(diagnostic['detail'], contains('actual_arity=${row['actual_arity']}'));
  expect(diagnostic['detail'], isNot(contains('must not run')));
}

Future<ProcessResult> _expectDartSuccess(
  Directory workingDirectory,
  Map<String, String> environment,
  List<String> arguments,
) async {
  final result = await Process.run(
    Platform.resolvedExecutable,
    arguments,
    workingDirectory: workingDirectory.path,
    environment: environment,
  );
  expect(
    result.exitCode,
    0,
    reason:
        'dart ${arguments.join(' ')} failed\n'
        'stdout:\n${result.stdout}\n'
        'stderr:\n${result.stderr}',
  );
  return result;
}

void main() {
  final contract = _contract();

  test('loads the exact logical-helper authority', () {
    expect(contract['format'], 1);
    expect(contract['contract_id'], 'linkedspec-logical-helper-v1');
    expect((contract['truthiness_cases']! as List), hasLength(17));
    expect((contract['helper_cases']! as List), hasLength(10));
  });

  test('every typed truth row uses the one Dart runtime seam', () {
    final rows = (contract['truthiness_cases']! as List)
        .cast<Map<Object?, Object?>>();

    for (final rawRow in rows) {
      final row = rawRow.cast<String, Object?>();
      final value = (row['value']! as Map).cast<String, Object?>();
      expect(
        runtimeLogicalTruth(_runtimeValue(value)),
        row['expected'],
        reason: row['id']! as String,
      );
    }
  });

  for (final fixtureId in ['values', 'effects', 'receiver_and_lazy_control']) {
    test('$fixtureId matches native and normalized execution', () {
      final fixture = _fixture(contract, fixtureId);
      final source = fixture['spec_source']! as String;
      final expected = fixture['expected'];
      final compiled = _compileSource(source);

      expect(
        LinkedSpecRuntimeEngine(compiled).parse('x').value,
        expected,
        reason: 'native $fixtureId',
      );

      final reconstructed = _reconstructFromEmittedPayload(
        compiled,
        'logical-helper/$fixtureId-normalized.spec',
      );
      expect(
        LinkedSpecRuntimeEngine(reconstructed).parse('x').value,
        expected,
        reason: 'normalized $fixtureId',
      );
    });

    test('$fixtureId matches generated-plan execution', () {
      final fixture = _fixture(contract, fixtureId);
      final source = fixture['spec_source']! as String;
      final compiled = _compileSource(source);

      final plan = buildGeneratedRulePlan(compiled);
      final identity = 'logical-helper/$fixtureId-generated.spec';
      final direct = executeGeneratedParserV1(compiled, plan, 'x', identity);
      expect(direct, fixture['expected']);
      expect(
        executeGeneratedParserWithTraceV1(
          compiled,
          plan,
          'x',
          LinkedSpecTraceConfig.disabled(),
          identity,
        ),
        direct,
      );
    });

    test('$fixtureId matches the primary Dart CLI', () {
      final fixture = _fixture(contract, fixtureId);
      final output = runLinkedSpecDartPrimaryCli([
        '--inline-spec',
        fixture['spec_source']! as String,
        '--input',
        'x',
      ]);

      expect(output.exitCode, 0);
      expect(output.stderrBytes, isEmpty);
      expect(jsonDecode(utf8.decode(output.stdoutBytes)), fixture['expected']);
    });
  }

  for (final fixture in _invalidFixtures(contract)) {
    final id = fixture['id']! as String;

    test('$id fails natively before operand effects', () {
      final row = _caseFor(_invalidCases(contract), id);
      final engine = LinkedSpecRuntimeEngine(
        _compileSource(fixture['spec_source']! as String),
      );

      try {
        engine.parse('x');
        fail('$id accepted invalid logical-helper arity');
      } on RuntimeInterpreterException catch (error) {
        _expectArityDiagnostic(error, row);
      } on Object catch (error) {
        fail('$id evaluated an operand or threw $error');
      }
    });

    test('$id fails through normalized and generated-plan execution', () {
      final row = _caseFor(_invalidCases(contract), id);
      final compiled = _compileSource(fixture['spec_source']! as String);
      final reconstructed = _reconstructFromEmittedPayload(
        compiled,
        'logical-helper/$id-normalized.spec',
      );

      try {
        LinkedSpecRuntimeEngine(reconstructed).parse('x');
        fail('$id normalized state accepted invalid arity');
      } on RuntimeInterpreterException catch (error) {
        _expectArityDiagnostic(error, row);
      } on Object catch (error) {
        fail('$id normalized state evaluated an operand or threw $error');
      }

      final identity = 'logical-helper/$id-generated.spec';
      final plan = buildGeneratedRulePlan(compiled);
      try {
        executeGeneratedParserV1(compiled, plan, 'x', identity);
        fail('$id generated plan accepted invalid arity');
      } on GeneratedSourceException catch (error) {
        _expectGeneratedArityDiagnostic(error, row, identity);
      } on Object catch (error) {
        fail('$id generated plan evaluated an operand or threw $error');
      }

      try {
        executeGeneratedParserWithTraceV1(
          compiled,
          plan,
          'x',
          LinkedSpecTraceConfig.disabled(),
          identity,
        );
        fail('$id traced generated plan accepted invalid arity');
      } on GeneratedSourceException catch (error) {
        _expectGeneratedArityDiagnostic(error, row, identity);
      } on Object catch (error) {
        fail('$id traced generated plan evaluated an operand or threw $error');
      }
    });

    test('$id keeps the primary Dart CLI failure projection', () {
      final output = runLinkedSpecDartPrimaryCli([
        '--inline-spec',
        fixture['spec_source']! as String,
        '--input',
        'x',
      ]);

      expect(output.exitCode, 1);
      expect(output.stdoutBytes, isEmpty);
      expect(
        output.stderrBytes,
        utf8.encode('linkedspec: parser invocation failed\n'),
      );
    });
  }

  test(
    'standalone emitted libraries preserve values effects controls and arity',
    () async {
      final packageRoot = Directory.current.absolute;
      final scratch = Directory.systemTemp.createTempSync(
        'linkedspec-dart-logical-helper-',
      );
      final pubCache = Directory('${scratch.path}/pub-cache')..createSync();

      try {
        final libraryDirectory = Directory('${scratch.path}/lib')..createSync();
        final binDirectory = Directory('${scratch.path}/bin')..createSync();
        File('${scratch.path}/pubspec.yaml').writeAsStringSync('''
name: linkedspec_logical_helper_probe
publish_to: none
environment:
  sdk: ">=3.9.0 <4.0.0"
dependencies:
  linkedspec_dart:
    path: ${jsonEncode(packageRoot.path)}
''');

        for (final fixtureId in [
          'values',
          'effects',
          'receiver_and_lazy_control',
        ]) {
          final source =
              _fixture(contract, fixtureId)['spec_source']! as String;
          File('${libraryDirectory.path}/$fixtureId.dart').writeAsStringSync(
            emitDartSourceV1(
              _compileSource(source),
              'logical-helper/$fixtureId-emitted.spec',
            ),
          );
        }
        final invalid = _invalidFixtures(
          contract,
        ).singleWhere((row) => row['id'] == 'not_many');
        File('${libraryDirectory.path}/invalid.dart').writeAsStringSync(
          emitDartSourceV1(
            _compileSource(invalid['spec_source']! as String),
            'logical-helper/not_many-emitted.spec',
          ),
        );
        File('${binDirectory.path}/main.dart').writeAsStringSync(r'''
import 'dart:convert';

import 'package:linkedspec_dart/linkedspec_dart.dart';
import 'package:linkedspec_logical_helper_probe/effects.dart' as effects;
import 'package:linkedspec_logical_helper_probe/invalid.dart' as invalid;
import 'package:linkedspec_logical_helper_probe/receiver_and_lazy_control.dart'
    as receiver;
import 'package:linkedspec_logical_helper_probe/values.dart' as values;

void main() {
  Object? invalidFailure;
  try {
    invalid.execute('x');
    throw StateError('not_many unexpectedly executed');
  } on GeneratedSourceException catch (error) {
    invalidFailure = error.toJson();
  }
  Object? invalidTracedFailure;
  try {
    invalid.executeWithTrace('x', LinkedSpecTraceConfig.disabled());
    throw StateError('traced not_many unexpectedly executed');
  } on GeneratedSourceException catch (error) {
    invalidTracedFailure = error.toJson();
  }
  print(jsonEncode({
    'values': {
      'direct': values.execute('x'),
      'traced': values.executeWithTrace('x', LinkedSpecTraceConfig.disabled()),
      'source_identity': values.metadata().sourceIdentity,
    },
    'effects': {
      'direct': effects.execute('x'),
      'traced': effects.executeWithTrace('x', LinkedSpecTraceConfig.disabled()),
      'source_identity': effects.metadata().sourceIdentity,
    },
    'receiver_and_lazy_control': {
      'direct': receiver.execute('x'),
      'traced': receiver.executeWithTrace('x', LinkedSpecTraceConfig.disabled()),
      'source_identity': receiver.metadata().sourceIdentity,
    },
    'invalid': invalidFailure,
    'invalid_traced': invalidTracedFailure,
  }));
}
''');

        final environment = {
          ...Platform.environment,
          'PUB_CACHE': pubCache.path,
        };
        await _expectDartSuccess(scratch, environment, const [
          'pub',
          'get',
          '--offline',
        ]);
        await _expectDartSuccess(scratch, environment, const [
          'analyze',
          '--fatal-infos',
          '--fatal-warnings',
        ]);
        final run = await _expectDartSuccess(scratch, environment, const [
          'run',
          'bin/main.dart',
        ]);
        final observed = (jsonDecode((run.stdout as String).trim()) as Map)
            .cast<String, Object?>();
        for (final fixtureId in [
          'values',
          'effects',
          'receiver_and_lazy_control',
        ]) {
          final roles = (observed[fixtureId]! as Map).cast<String, Object?>();
          expect(
            roles['direct'],
            _fixture(contract, fixtureId)['expected'],
            reason: '$fixtureId direct',
          );
          expect(roles['traced'], roles['direct'], reason: '$fixtureId traced');
          expect(
            roles['source_identity'],
            'logical-helper/$fixtureId-emitted.spec',
          );
        }
        final failure = (observed['invalid']! as Map).cast<String, Object?>();
        expect(failure['stage'], 'execute_generated');
        expect(failure['code'], 'generated_execution_failed');
        expect(failure['detail'], contains('helper_arity_mismatch'));
        expect(failure['detail'], contains('helper_name=not'));
        expect(failure['detail'], contains('actual_arity=2'));
        expect(failure['detail'], isNot(contains('must not run')));
        expect(
          failure['source_identity'],
          'logical-helper/not_many-emitted.spec',
        );
        final tracedFailure = (observed['invalid_traced']! as Map)
            .cast<String, Object?>();
        expect(tracedFailure, failure);
      } finally {
        scratch.deleteSync(recursive: true);
      }

      expect(scratch.existsSync(), isFalse);
    },
    timeout: const Timeout(Duration(minutes: 2)),
  );
}

File: lua/test/logical_helper_contract_test.lua
local linkedspec = require("linkedspec")
local json = linkedspec.json

local assertions = 0
local failures = {}

local function check(condition, label)
  assertions = assertions + 1
  if not condition then failures[#failures + 1] = label or "assertion failed" end
end

local function check_equal(actual, expected, label)
  check(actual == expected, (label or "values differ") ..
    ": expected " .. tostring(expected) .. ", got " .. tostring(actual))
end

local function check_contains(actual, expected, label)
  check(tostring(actual):find(expected, 1, true) ~= nil, (label or "text differs") ..
    ": expected to contain " .. expected .. ", got " .. tostring(actual))
end

local function check_same_json(actual, expected, label)
  check_equal(json.encode(actual), json.encode(expected), label)
end

local function read_file(path)
  local handle = assert(io.open(path, "rb"))
  local value = assert(handle:read("*a"))
  assert(handle:close())
  return value
end

local contract = json.decode(read_file("capability_conformance/logical_helper_contract.json"))

local function contract_row(rows, id)
  for _, row in ipairs(rows) do
    if row.id == id then return row end
  end
  error("logical-helper contract row is missing: " .. id, 0)
end

local function compile_source(source)
  return linkedspec.compile_spec(linkedspec.parse_spec(source))
end

local function reconstruct_source(source)
  local parsed = linkedspec.parse_spec(source)
  local serialized = json.encode(linkedspec.spec_ast.to_json(parsed))
  return linkedspec.compile_spec(
    linkedspec.spec_ast.from_json("SpecFile", json.decode(serialized))
  )
end

local function execute_native(compiled)
  return linkedspec.runtime_parse(linkedspec.runtime_engine(compiled), "x").value
end

local function generated_module(source, name)
  local loader = loadstring or load
  local chunk, failure = loader(source, name)
  if chunk == nil then error(failure, 0) end
  return chunk()
end

local function generated_identity(id, role)
  return "logical-helper/" .. id .. "-" .. role .. ".spec"
end

local function execute_generated(compiled, id)
  return linkedspec.execute_generated_parser_v1(
    compiled,
    linkedspec.build_generated_rule_plan(compiled),
    "x",
    generated_identity(id, "generated")
  )
end

local function emitted_module(compiled, id)
  return generated_module(
    linkedspec.emit_lua_source_v1(compiled, generated_identity(id, "emitted")),
    "@logical-helper-" .. id .. "-emitted"
  )
end

local function primary(source)
  return linkedspec.run_primary_cli({ "--inline-spec", source, "--input", "x" })
end

local function capture(operation)
  local ok, value = pcall(operation)
  return ok, value
end

local function check_runtime_arity_failure(ok, failure, row, label)
  check_equal(ok, false, label .. " rejected")
  local typed = not ok and linkedspec.is_runtime_interpreter_error(failure)
  check_equal(typed, true, label .. " typed runtime failure")
  local fields = typed and failure or {}
  check_equal(fields.code, row.expected_code, label .. " code")
  check_equal(fields.helper_name, row.helper_name, label .. " helper name")
  check_equal(fields.actual_arity, row.actual_arity, label .. " actual arity")
  check_equal(fields.expected_arity, row.expected_arity, label .. " expected arity")
  check_contains(fields.message or "", "helper_arity_mismatch", label .. " code token")
  check_contains(fields.message or "", "helper_name=" .. row.helper_name, label .. " helper token")
  check_contains(fields.message or "", "actual_arity=" .. row.actual_arity, label .. " actual token")
  check_equal((fields.message or ""):find("must not run", 1, true), nil, label .. " no operand effect")

  local diagnostic = typed and fields.diagnostic or nil
  check_equal(linkedspec.is_runtime_diagnostic(diagnostic), true, label .. " diagnostic type")
  local diagnostic_json = diagnostic and linkedspec.interpreter.to_json(diagnostic) or {}
  check_equal(diagnostic_json.stage, "helper_arity_mismatch", label .. " diagnostic stage")
  check_equal(diagnostic_json.code, row.expected_code, label .. " diagnostic code")
  check_equal(diagnostic_json.helper_name, row.helper_name, label .. " diagnostic helper")
  check_equal(diagnostic_json.actual_arity, row.actual_arity, label .. " diagnostic actual")
  check_equal(diagnostic_json.expected_arity, row.expected_arity, label .. " diagnostic expected")
  check_equal(diagnostic_json.rule_label, "Top", label .. " diagnostic rule")
end

local function check_generated_arity_failure(ok, failure, row, label)
  check_equal(ok, false, label .. " rejected")
  local typed = not ok and linkedspec.is_generated_source_error(failure)
  check_equal(typed, true, label .. " typed generated failure")
  local fields = typed and failure or {}
  check_equal(
    fields.stage and linkedspec.generated_source_stage_name(fields.stage),
    "execute_generated",
    label .. " stage"
  )
  check_equal(
    fields.code and linkedspec.generated_source_code_name(fields.code),
    "generated_execution_failed",
    label .. " code"
  )
  local detail = fields.detail or ""
  check_contains(detail, "helper_arity_mismatch", label .. " detail code")
  check_contains(detail, "helper_name=" .. row.helper_name, label .. " detail helper")
  check_contains(detail, "actual_arity=" .. row.actual_arity, label .. " detail actual")
  check_equal(detail:find("must not run", 1, true), nil, label .. " no operand effect")
end

check_equal(contract.format, 1, "contract format")
check_equal(contract.contract_id, "linkedspec-logical-helper-v1", "contract id")
check_equal(#contract.truthiness_cases, 17, "truthiness case count")
check_equal(#contract.helper_cases, 10, "helper case count")

do
  local codeblock_source = table.concat({
    "fn truth(callback: codeblock) { return(and(callback)) }",
    "Top::",
    " /x/",
    ' E { return(truth({ fail("codeblock must not run") })) }',
    "",
  }, "\n")
  local parsed = linkedspec.parse_spec_with_staged_user_function_definitions(codeblock_source)
  local result = execute_native(linkedspec.compile_spec(parsed))
  check_equal(result, true, "inert codeblock is truthful without invocation")
end

for _, fixture_id in ipairs({ "values", "effects", "receiver_and_lazy_control" }) do
  local fixture = contract.fixtures[fixture_id]
  local compiled = compile_source(fixture.spec_source)
  check_same_json(execute_native(compiled), fixture.expected, fixture_id .. " native")
  check_same_json(
    execute_native(reconstruct_source(fixture.spec_source)),
    fixture.expected,
    fixture_id .. " reconstructed"
  )
  check_same_json(execute_generated(compiled, fixture_id), fixture.expected, fixture_id .. " generated")
  check_same_json(emitted_module(compiled, fixture_id).execute("x"), fixture.expected, fixture_id .. " emitted")

  local primary_result = primary(fixture.spec_source)
  check_equal(primary_result.exit_code, 0, fixture_id .. " primary exit")
  check_equal(primary_result.stderr, "", fixture_id .. " primary stderr")
  check_same_json(json.decode(primary_result.stdout), fixture.expected, fixture_id .. " primary value")
end

for _, fixture in ipairs(contract.fixtures.invalid_arity) do
  local row = contract_row(contract.invalid_arity_cases, fixture.id)
  local compiled = compile_source(fixture.spec_source)
  local native_ok, native_failure = capture(function() return execute_native(compiled) end)
  check_runtime_arity_failure(native_ok, native_failure, row, fixture.id .. " native")

  local reconstructed = reconstruct_source(fixture.spec_source)
  local reconstructed_ok, reconstructed_failure = capture(function()
    return execute_native(reconstructed)
  end)
  check_runtime_arity_failure(
    reconstructed_ok,
    reconstructed_failure,
    row,
    fixture.id .. " reconstructed"
  )

  local generated_ok, generated_failure = capture(function()
    return execute_generated(compiled, fixture.id)
  end)
  check_generated_arity_failure(generated_ok, generated_failure, row, fixture.id .. " generated")

  local emitted = emitted_module(compiled, fixture.id)
  local emitted_ok, emitted_failure = capture(function() return emitted.execute("x") end)
  check_generated_arity_failure(emitted_ok, emitted_failure, row, fixture.id .. " emitted")

  local primary_result = primary(fixture.spec_source)
  check_equal(primary_result.exit_code, 1, fixture.id .. " primary exit")
  check_equal(primary_result.stdout, "", fixture.id .. " primary stdout")
  check_equal(primary_result.stderr, "linkedspec: parser invocation failed\n", fixture.id .. " primary stderr")
end

if #failures == 0 then
  io.stdout:write("logical helper contract: ", assertions, " assertions passed\n")
else
  io.stderr:write("logical helper contract: ", #failures, " of ", assertions, " assertions failed\n")
  for _, message in ipairs(failures) do io.stderr:write("- ", message, "\n") end
  os.exit(1)
end
